lib/recursive-methods.rb

```diff
@@ -1,49 +1,98 @@
 # Authoring recursive algorithms. Add comments including time and space complexity for each method.
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def factorial(n)
-    raise NotImplementedError, "Method not implemented"
+    #base
+    return 1 if n == 0
+    raise ArgumentError if n < 0
+    #recursion
+    return n * factorial(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n^2)
+# Space complexity: O(n)
 def reverse(s)
-    raise NotImplementedError, "Method not implemented"
+    return s if s.length == 1 || s.length == 0
+    return reverse_helper(s,first = 0, last = s.length - 1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+def reverse_helper(s, first, last)
+    if first < last
+        s[first],s[last] = s[last],s[first]
+        return reverse_helper(s, first+1, last-1)
+    end
+    return s
+end
+
+# Time complexity: O(n^2)
+# Space complexity: O(n)
 def reverse_inplace(s)
-    raise NotImplementedError, "Method not implemented"
+    return reverse_helper(s, 0, s.length-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n)
 def bunny(n)
-    raise NotImplementedError, "Method not implemented"
+    #base
+    return 0 if n == 0
+    #recursion
+    return 2 + bunny(n-1)
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2) because s[1..-2] creates a new string with each recursion
 def nested(s)
-    raise NotImplementedError, "Method not implemented"
+    return true if s.length == 0
+    return false if s.length.odd?
+
+    # s.count("(") == s.count(")")
+    return (s[0] === "(" && s[-1] === ")") && nested(s[1..-2]);
 end
 
-# Time complexity: ?
-# Space complexity: ?
+# Time complexity: O(n)
+# Space complexity: O(n^2)
 def search(array, value)
```

Comment:
The space complexity is O(n) because you're not making a new array with each method call.